Data/Aeson/Types/FromJSON.hs

```diff
@@ -1273,38 +1274,71 @@ instance (ProductFromJSON arity f, ProductSize f
 
 --------------------------------------------------------------------------------
 
+class FieldNames f where
+    fieldNames :: f a -> [Text]
+
+instance (FieldNames a, FieldNames b) => FieldNames (a :*: b) where
+    fieldNames _ =
+      fieldNames (undefined :: a x) ++
```

Comment:
Use a difference list here (if we are to use set difference instead of list difference `\\` below).

```haskell
    fieldNames :: f a -> [Text] -> [Text]
```

```haskell
    fieldNames _ =
      fieldNames (undefined :: a x) .
      fieldNames (undefined :: b x)
```

```haskell
    fieldNames _ = (pack (selName (undefined M1 _i s _f _p)) :)
```